include/seqan3/alignment/matrix/debug_matrix.hpp

```diff
@@ -0,0 +1,391 @@
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ * \brief Contains the declaration of seqan3::detail::debug_matrix.
+ */
+
+#pragma once
+
+#include <iomanip>
+
+#include <seqan3/alignment/matrix/matrix_concept.hpp>
+#include <seqan3/alignment/matrix/row_wise_matrix.hpp>
+#include <seqan3/alignment/matrix/trace_directions.hpp>
+#include <seqan3/core/debug_stream.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief A debug matrix to wrap alignment matrices and sequences and make them printable together.
+ * \ingroup alignment_matrix
+ * \implements seqan3::detail::Matrix
+ * \tparam matrix_t    An alignment matrix; Must model seqan3::detail::Matrix.
+ * \tparam sequence1_t The type of the first sequence; If no sequences are given this is std::nullopt_t.
+ * \tparam sequence2_t The type of the second sequence; If no sequences are given this is std::nullopt_t.
+ *
+ *
+ * This debug matrix allows you to print an alignment matrix (e.g. score or trace matrix) combined with two sequences.
+ *
+ * \cond DEV
+ * This type is used internally
+ *   * to print seqan3::detail::Matrix's and,
+ *   * to compare alignment matrices in in test cases.
+ * \endcond
+ *
+ * # Score Matrix Example
+ *
+ * \snippet test/snippet/alignment/matrix/debug_matrix.cpp score_matrix
+ *
+ * ### Output
+ * \snippet test/snippet/alignment/matrix/debug_matrix.out score_matrix::out
+ *
+ * # Trace Matrix Example
+ *
+ * \snippet test/snippet/alignment/matrix/debug_matrix.cpp trace_matrix
+ *
+ * ### Output
+ * \snippet test/snippet/alignment/matrix/debug_matrix.out trace_matrix::out
+ */
+template <Matrix matrix_t, typename sequence1_t = std::nullopt_t, typename sequence2_t = std::nullopt_t>
+class debug_matrix
+{
+public:
+    //!\copydoc seqan3::detail::Matrix::entry_type
+    using entry_type = typename std::remove_reference_t<matrix_t>::entry_type;
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    debug_matrix() = default; //!< Defaulted
+    debug_matrix(debug_matrix const &) = default; //!< Defaulted
+    debug_matrix(debug_matrix &&) = default; //!< Defaulted
+    debug_matrix & operator=(debug_matrix const &) = default; //!< Defaulted
+    debug_matrix & operator=(debug_matrix &&) = default; //!< Defaulted
+    ~debug_matrix() = default;  //!< Defaulted
+
+    /*!\brief Construct the matrix out of the *entries*, the *rows*, and the *cols*.
+     * \param entries The entry values as a flat std::vector <#entry_type>.
+     * \param rows    The number of rows.
+     * \param cols    The number of columns.
+     */
+    debug_matrix(std::vector<entry_type> entries, size_t const rows, size_t const cols)
```

Comment:
I personally would remove this constructor.
Obviously I can call already `debug_matrix{matrix_t{entries, rows, cols}}`;
In this case the matrix details are not coupled with the debug_matrix.